db/db_test.go

```diff
@@ -111,4 +141,197 @@ func TestLoadBadSpec(t *testing.T) {
+	assert.False(ok)
+
+	act, err = tx.Get("bar")
+	assert.NoError(err)
+	assert.Nil(act)
+
+	err = tx.Close()
+	assert.NoError(err)
+}
+
+func TestClosedTransaction(t *testing.T) {
+	assert := assert.New(t)
+	db, _ := LoadTempDB(assert)
+
+	tx := db.NewTransaction()
+	assert.False(tx.Closed())
+	_, err := tx.Commit()
+	assert.NoError(err)
+	assert.True(tx.Closed())
+
+	_, err = tx.Has("k")
+	assert.Equal(ClosedError, err)
+	_, err = tx.Get("k")
+	assert.Equal(ClosedError, err)
+	_, err = tx.Scan(ScanOptions{})
+	assert.Equal(ClosedError, err)
+	err = tx.Put("k", []byte(`"v"`))
+	assert.Equal(ClosedError, err)
+	_, err = tx.Del("k")
+	assert.Equal(ClosedError, err)
+	_, err = tx.Commit()
+	assert.Equal(ClosedError, err)
+	err = tx.Close()
+	assert.Equal(ClosedError, err)
+}
+
+func TestWriteTransaction(t *testing.T) {
+	assert := assert.New(t)
+	db, _ := LoadTempDB(assert)
+
+	exp := []byte(`"bar"`)
+	tx := db.NewTransaction()
+
+	err := tx.Put("foo", exp)
+	assert.NoError(err)
+
+	ok, err := tx.Has("foo")
+	assert.NoError(err)
+	assert.True(ok)
+	act, err := tx.Get("foo")
+	assert.NoError(err)
+	assert.Equal(exp, act, "expected %s got %s", exp, act)
+
+	ok, err = tx.Has("bar")
+	assert.NoError(err)
+	assert.False(ok)
+
+	act, err = tx.Get("bar")
+	assert.NoError(err)
+	assert.Nil(act)
+
+	assertDataEquals(assert, db, `map {}`)
+	_, err = tx.Commit()
+	assert.NoError(err)
+	assertDataEquals(assert, db, `map {"foo": "bar"}`)
+}
+
+func TestReadAndWriteTransaction(t *testing.T) {
+	assert := assert.New(t)
+	db, _ := LoadTempDB(assert)
+
+	exp := []byte(`"bar"`)
+	rtx := db.NewTransaction()
+
+	wtx := db.NewTransaction()
+	err := wtx.Put("foo", exp)
+	assert.NoError(err)
+
+	has, err := wtx.Has("foo")
+	assert.NoError(err)
+	assert.True(has)
+
+	has, err = rtx.Has("foo")
+	assert.NoError(err)
+	assert.False(has)
+
+	act, err := wtx.Get("foo")
+	assert.NoError(err)
+	assert.Equal(exp, act, "expected %s got %s", exp, act)
+
+	act, err = rtx.Get("foo")
+	assert.NoError(err)
+	assert.Nil(act)
+
+	assertDataEquals(assert, db, `map {}`)
+
+	_, err = wtx.Commit()
+	assert.NoError(err)
+
+	has, err = rtx.Has("foo")
+	assert.NoError(err)
+	assert.False(has, "Read transaction should still operate at old head")
+
+	err = rtx.Close()
+	assert.NoError(err)
+}
+
+func TestMultipleWriteTransaction(t *testing.T) {
+	assert := assert.New(t)
+	db, _ := LoadTempDB(assert)
+
+	v1 := []byte(`"v1"`)
+	v2 := []byte(`"v2"`)
+
+	tx1 := db.NewTransaction()
+	err := tx1.Put("k", v1)
+	assert.NoError(err)
+
+	tx2 := db.NewTransaction()
+	err = tx2.Put("k", v2)
+	assert.NoError(err)
+
+	act, err := tx1.Get("k")
+	assert.NoError(err)
+	assert.Equal(v1, act, "expected %s got %s", v1, act)
+
+	act, err = tx2.Get("k")
+	assert.NoError(err)
+	assert.Equal(v2, act, "expected %s got %s", v2, act)
+
+	_, err = tx1.Commit()
+	assert.NoError(err)
+	assertDataEquals(assert, db, `map {"k": "v1"}`)
+
+	_, err = tx2.Commit()
+	assert.Equal("Dataset head is not ancestor of commit", err.Error())
```

Comment:
Can we detect this and have a more clear error, particularly since callee is going to have to retry? (Can be a separate PR).